### ISSUE TITLE
feat(interpolate): KDE points-to-grid (kde_to_grid via sklearn.neighbors.KernelDensity)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/xrtoolz#181]
# Design: KDE points-to-grid

**Status:** design draft (MVP)
**Author:** @jejjohnson
**Date:** 2026-05-07 (updated 2026-05-09: switched backend scipy→sklearn)
**Backend:** `sklearn.neighbors.KernelDensity` (already a hard dep — see [`pyproject.toml`](../pyproject.toml) `scikit-learn>=1.4`)
**Module home:** extension of [`src/xr_toolz/interpolate/_src/points_to_grid.py`](../src/xr_toolz/interpolate/_src/points_to_grid.py) + operator in [`interpolate/operators.py`](../src/xr_toolz/interpolate/operators.py)
**Related:** existing `points_to_grid` (binned histograms); kNN/IDW issue (interpolation, not density); [#158](https://github.com/jejjohnson/xr_toolz/issues/158) (dask)

---

## 1. User story

> As a user with **sparse / unevenly distributed observations** (AIS ship tracks, drifter trajectories, sparse buoys, weather-station locations), I want a **smooth density map** on a regular grid — not a histogram with empty cells. Histograms are noisy with small N; KDE smooths the density estimate using a kernel and produces a continuous field.

`points_to_grid` today does histogram binning (`scipy.stats.binned_statistic_2d`). For density questions ("where do drifters spend their time?", "what's the AIS shipping density?"), we want KDE.

## 2. Why sklearn.neighbors.KernelDensity (and not scipy)

| Option | Verdict |
|---|---|
| **`sklearn.neighbors.KernelDensity`** | **Pick.** scikit-learn is already a hard dep (`pyproject.toml`). Tree-backed (`kd_tree` / `ball_tree`) — scales to N ≫ 10⁵ via radius cutoffs; six built-in kernels (gaussian, tophat, epanechnikov, exponential, linear, cosine); supports `metric="haversine"` for great-circle KDE on the sphere — directly relevant to global geoscience density. |
| `scipy.stats.gaussian_kde` | O(N × M) brute force; Gaussian-only; Euclidean only. Adaptive bandwidth via Scott / Silverman rules built-in (we re-implement these for sklearn — ~5 lines). Fine for small N but doesn't scale and forces planar geometry. |
| Hand-rolled FFT-KDE | Faster for very large grids but adds maintenance burden and no haversine support. |

**Geoscience-specific wins from sklearn:**

1. **Tree-based local density.** `kd_tree` / `ball_tree` make nearest- and radius-restricted estimators tractable on AIS-scale point clouds (N ~ 10⁶+). With `rtol > 0`, evaluation prunes far-away points and the cost goes from O(N × M) to roughly O((N + M) log N).
2. **Spherical KDE.** `metric="haversine"` evaluates directly on the sphere, eliminating the pole bias that comes from treating (lon, lat) as planar. This is the single most-requested KDE feature for global geoscience density (shipping, drifter tracks, lightning, …).
3. **Kernel choice.** Compact-support kernels (tophat, epanechnikov, linear, cosine) are useful when users want a "soft histogram" with a hard radius. Gaussian alone forces a long tail.
4. **Already in our stack.** Adopting sklearn here is consistent with [#177](https://github.com/jejjohnson/xr_toolz/issues/177) (kNN/IDW) and [#178](https://github.com/jejjohnson/xr_toolz/issues/178) (GP), which are also sklearn-backed.

## 3. Scope (MVP)

In:
- **`kde_to_grid`** — scattered (lon, lat[, weights]) → density grid via sklearn `KernelDensity`.
- **`KDEToGrid`** operator.
- **Bandwidth selection:** `bandwidth: float | "scott" | "silverman" = "scott"`. Float = explicit; "scott" / "silverman" computed in-house from sample size + covariance trace (`h = N^{-1/(d+4)} · σ̂` per scipy convention) so users get the same defaults as the scipy version.
- **Kernel choice:** `kernel: Literal["gaussian", "tophat", "epanechnikov", "exponential", "linear", "cosine"] = "gaussian"`.
- **Metric:** `metric: Literal["euclidean", "haversine"] = "euclidean"`. `"haversine"` evaluates on the sphere (inputs/bandwidth in radians; we convert lon/lat→radians internally and document the bandwidth-units gotcha).
- **Tree backend:** `algorithm: Literal["auto", "kd_tree", "ball_tree"] = "auto"` (`ball_tree` is required for haversine; we auto-select).
- **Output normalization:** `output: "density" | "counts" | "counts_per_area" = "density"`.
- **Tree pruning tolerance:** `rtol: float = 1e-4`. sklearn's default of 1e-8 is unnecessarily strict for visualization-grade density.

Out (file as follow-ups):
- 3-D / spatio-temporal KDE (`(time, lat, lon)`).
- Adaptive (per-point) bandwidth.
- FFT-accelerated KDE for very large grids (e.g. `KDEpy.FFTKDE`) — sklearn already gets us tree-acceleration; FFT only helps when M is huge and the kernel is separable.
- Cross-validated bandwidth selection (`GridSearchCV` over `KernelDensity`) — easy follow-up since we're already on sklearn.

## 4. Mathematics

For points $\{\mathbf{x}_i\}_{i=1}^N$ with optional weights $w_i$ ($\sum w_i = 1$, default $w_i = 1/N$), the KDE estimate at $\mathbf{x}^*$ is

$$
\hat p(\mathbf{x}^*) = \sum_{i=1}^N w_i\, K_h(\mathbf{x}^* - \mathbf{x}_i),
$$

where $K_h$ is one of sklearn's six kernels with bandwidth $h$. For the Gaussian kernel,

$$
K_h(\mathbf{u}) = (2\pi h^2)^{-d/2} \exp\!\left(-\frac{\|\mathbf{u}\|^2}{2 h^2}\right).
$$

**Bandwidth rules (computed manually to match scipy semantics).** Let $\hat\sigma = \sqrt{\mathrm{tr}(\hat\Sigma)/d}$ be the mean-coordinate sample standard deviation.

- Scott: $h = N^{-1/(d+4)} \cdot \hat\sigma$.
- Silverman: $h = (N(d+2)/4)^{-1/(d+4)} \cdot \hat\sigma$.

For `metric="haversine"` we compute $\hat\sigma$ on the radian inputs, so the rule produces a bandwidth in radians.

**Output modes:**

- `"density"` — values integrate to 1 over the grid (per area). Units: [lon × lat]⁻¹ for a Euclidean (lon, lat) grid; rad⁻² for haversine.
- `"counts"` — density × N. Approximate "expected count per cell area".
- `"counts_per_area"` — counts × cell area. Approximate "count per square degree" (or per cell).

### 5.1 Layer 0 — `_src/points_to_grid.py` (extension)

```python
def kde_to_grid(
    lons: ArrayLike,
    lats: ArrayLike,
    grid: Grid,
    *,
    weights: ArrayLike | None = None,
    bandwidth: float | Literal["scott", "silverman"] = "scott",
    kernel: Literal[
        "gaussian", "tophat", "epanechnikov", "exponential", "linear", "cosine"
    ] = "gaussian",
    metric: Literal["euclidean", "haversine"] = "euclidean",
    algorithm: Literal["auto", "kd_tree", "ball_tree"] = "auto",
    output: Literal["density", "counts", "counts_per_area"] = "density",
    rtol: float = 1e-4,
) -> xr.DataArray:
    """KDE of (lon, lat) points evaluated on ``grid`` via sklearn.

    Args:
        lons, lats: 1-D point coordinates (NaN entries dropped).
        grid: Target :class:`Grid`.
        weights: Optional per-point weights (passed to ``KernelDensity.fit``
            as ``sample_weight``; non-finite entries dropped).
        bandwidth: Float, or "scott" / "silverman" rules-of-thumb. For
            ``metric="haversine"`` a float is interpreted in radians.
        kernel: Any sklearn-supported kernel.
        metric: "euclidean" treats (lon, lat) as planar; "haversine" computes
            great-circle KDE on the sphere (inputs converted to radians
            internally; bandwidth must be in radians).
        algorithm: Tree backend. Auto-promoted to "ball_tree" when
            metric="haversine".
        output: Density (integrates to 1), total counts, or counts per area.
        rtol: Relative tolerance for tree pruning. sklearn's default 1e-8 is
            slow; 1e-4 is plenty for visualization-grade density.

    Returns:
        DataArray on ``grid`` with the chosen normalization.
    """
```

### 5.2 Layer 1 — `interpolate/operators.py`

```python
class KDEToGrid(Operator):
    """Wrap :func:`kde_to_grid`. Input: ``(lons, lats[, weights])``."""

    def __init__(
        self,
        grid: _binning.Grid,
        *,
        bandwidth: float | str = "scott",
        kernel: str = "gaussian",
        metric: str = "euclidean",
        algorithm: str = "auto",
        output: str = "density",
        rtol: float = 1e-4,
    ): ...
```

### 5.3 Public surface

Re-export `kde_to_grid` and `KDEToGrid` from `interpolate/__init__.py`. Group with `points_to_grid` / `PointsToGrid` in the docstring nav.

## 6. Implementation sketch

```python
# _src/points_to_grid.py (additions)
import numpy as np
import xarray as xr
from sklearn.neighbors import KernelDensity
from xr_toolz.interpolate._src.binning import Grid


def _bandwidth_rule(pts: np.ndarray, rule: str) -> float:
    # pts: (N, d). Match scipy's scott/silverman semantics.
    n, d = pts.shape
    sigma = float(np.sqrt(np.trace(np.cov(pts, rowvar=False)) / d))
    if rule == "scott":
        return n ** (-1.0 / (d + 4)) * sigma
    if rule == "silverman":
        return (n * (d + 2) / 4.0) ** (-1.0 / (d + 4)) * sigma
    raise ValueError(f"unknown bandwidth rule {rule!r}")


def kde_to_grid(
    lons, lats, grid, *, weights=None,
    bandwidth="scott", kernel="gaussian",
    metric="euclidean", algorithm="auto",
    output="density", rtol=1e-4,
):
    if output not in ("density", "counts", "counts_per_area"):
        raise ValueError(f"unknown output mode {output!r}")
    lons = np.ravel(np.asarray(lons))
    lats = np.ravel(np.asarray(lats))
    finite = np.isfinite(lons) & np.isfinite(lats)
    if weights is not None:
        w = np.ravel(np.asarray(weights, dtype=float))
        finite &= np.isfinite(w)
        w = w[finite]
    else:
        w = None
    pts = np.column_stack([lons[finite], lats[finite]])  # (N, 2)
    if pts.shape[0] < 2:
        raise ValueError("KDE requires at least 2 finite points")

    # Haversine wants (lat, lon) in radians; ball_tree only.
    if metric == "haversine":
        fit_pts = np.deg2rad(pts[:, [1, 0]])  # (lat, lon) rad
        algorithm = "ball_tree"
    else:
        fit_pts = pts

    if isinstance(bandwidth, str):
        h = _bandwidth_rule(fit_pts, bandwidth)
    else:
        h = float(bandwidth)

    kde = KernelDensity(
        bandwidth=h, kernel=kernel, metric=metric,
        algorithm=algorithm, rtol=rtol,
    ).fit(fit_pts, sample_weight=w)

    lon_g, lat_g = np.meshgrid(grid.lon, grid.lat, indexing="xy")
    if metric == "haversine":
        queries = np.deg2rad(np.column_stack([lat_g.ravel(), lon_g.ravel()]))
    else:
        queries = np.column_stack([lon_g.ravel(), lat_g.ravel()])
    log_density = kde.score_samples(queries)
    density = np.exp(log_density).reshape(len(grid.lat), len(grid.lon))

    n_eff = pts.shape[0] if w is None else float(w.sum())
    if output == "density":
        out = density
    elif output == "counts":
        out = density * n_eff
    else:  # counts_per_area
        dlon = float(np.mean(np.diff(grid.lon)))
        dlat = float(np.mean(np.diff(grid.lat)))
        out = density * n_eff * dlon * dlat
    return xr.DataArray(
        out, dims=("lat", "lon"),
        coords={"lat": grid.lat, "lon": grid.lon},
    )
```

## 7. Usage examples

```python
from xr_toolz.interpolate import kde_to_grid, KDEToGrid
from xr_toolz.interpolate._src.binning import Grid

# Drifter density on a 1° grid (planar)
grid = Grid(lon=np.arange(-180, 180, 1.0), lat=np.arange(-80, 80, 1.0))
density = kde_to_grid(drifter_lons, drifter_lats, grid, output="density")
density.plot()

# Global AIS density on the sphere (no pole bias)
density = kde_to_grid(
    ship_lons, ship_lats, grid,
    metric="haversine", bandwidth=0.05,    # rad ≈ 290 km
    kernel="epanechnikov",                  # compact support — soft histogram
    output="counts_per_area",
)

# Counts-per-cell weighted by track duration
density = kde_to_grid(
    drifter_lons, drifter_lats, grid,
    weights=track_durations, output="counts",
)
```

## 8. Tests (DoD)

1. **Single point.** A single point at the grid centre with `bandwidth=1.0`, Gaussian kernel, produces a density centred there; integral over grid ≈ 1 for `output="density"`.
2. **Density integrates to 1.** For any kernel, `(density * cell_area).sum()` ≈ 1 within bandwidth-truncation error.
3. **Counts mode preserves N.** `output="counts"` summed over the grid times cell area ≈ N.
4. **Bandwidth dispatch.** `0.5`, `"scott"`, `"silverman"` all run and produce different finite outputs.
5. **Kernel dispatch.** All six sklearn kernels run; tophat/epanechnikov produce zero outside their support radius.
6. **Haversine metric.** Two points 180° apart on the equator vs near a pole give similar density (Euclidean would not — that's the test of the win).
7. **Weights honored.** Equal weights match unweighted call. Concentrated weights peak where the weighted points are.
8. **NaN-input handling.** Non-finite (lon, lat) entries dropped; non-finite weights also dropped.
9. **Too-few-points guard.** N < 2 raises `ValueError`.
10. **Operator round-trip.** `KDEToGrid` config carries `<Grid>` placeholder per existing convention.

## 9. Subtasks

1. Implement `kde_to_grid` per §6 (sklearn-backed).
2. Bandwidth-rule helper (`_bandwidth_rule`) matching scipy semantics.
3. Operator + public exports.
4. Tests per §8 — including a haversine-vs-euclidean comparison test that demonstrates the pole-bias fix.
5. Docs notebook: AIS / drifter density example with binned-histogram baseline side-by-side, plus a global-haversine example demonstrating the spherical KDE benefit.

## 10. Open questions

- **Default `bandwidth`.** Scott's rule is the historical default and a sensible starting point. Silverman is similar; both can over-smooth on bimodal data. **Recommendation:** `"scott"`.
- **Default `output` mode.** "density" is the mathematical default and integrates to 1 — easiest to interpret. **Recommendation:** `"density"`.
- **Default `kernel`.** Gaussian for parity with the previous scipy design and because it has infinite support (no zero cells). **Recommendation:** `"gaussian"`.
- **Default `metric`.** "euclidean" preserves backwards-compat with how users already think about (lon, lat) grids; haversine is opt-in. **Recommendation:** `"euclidean"`, document haversine as the right call for global density.
- **Bandwidth units in haversine mode.** sklearn's haversine expects radians, so a float bandwidth must also be radians. Document with a "≈ km / R_earth" rule of thumb in the docstring.
- **Performance ceiling.** sklearn's tree-based KDE is O((N + M) log N) effective; comfortably handles N ~ 10⁶ on a 1° global grid with `rtol=1e-4`. FFT-KDE only worth adding if users hit the wall.
- **Spherical KDE.** Solved by `metric="haversine"`. **Document.**
- **Per-cell bandwidth.** Adaptive KDE is a real improvement for very-non-uniform point distributions. **Punt** until asked — sklearn doesn't support it natively.
- **Cross-validated bandwidth.** Trivial to add via `GridSearchCV(KernelDensity(), {"bandwidth": ...})`. **Punt** to follow-up unless asked.
